spec/acceptance/realtime/message_spec.rb

```diff
@@ -736,4 +736,128 @@ def self.add_tests_for_data(data)
       end
     end
   end
+
+  def expect_data_to_match(spec_value, message_data)
+    case message_data
+    when Hash
+      # Ignore ordering
+      expect(spec_value).to include(message_data)
+    else
+      expect(message_data).to eql(spec_value)
+    end
+  end
+
+  context 'message encoding interoperability' do
+    let(:client_options)  { { key: api_key, environment: environment, protocol: :json } }
+    let(:channel_name) { "subscribe_send_text-#{random_str}" }
+
+    fixtures_path = File.expand_path('../../../../lib/submodules/ably-common/test-resources/messages-encoding.json', __FILE__)
+
+    context 'over a JSON transport' do
+      let(:realtime_client) do
+        auto_close Ably::Realtime::Client.new(client_options)
+      end
+      let(:rest_client) do
+        Ably::Rest::Client.new(client_options)
+      end
+      let(:realtime_channel) { realtime_client.channels.get(channel_name) }
+
+      JSON.parse(File.read(fixtures_path))['messages'].each do |encoding_spec|
+        context "when decoding #{encoding_spec['expectedType']}" do
+          it 'ensures that client libraries have compatible encoding and decoding using common fixtures' do
+            realtime_channel.attach do
+              realtime_channel.subscribe do |message|
+                if encoding_spec['expectedHexValue']
+                  expect(message.data.unpack('H*').first).to eql(encoding_spec['expectedHexValue'])
+                else
+                  expect_data_to_match encoding_spec['expectedValue'], message.data
+                end
+                stop_reactor
+              end
+
+              raw_message = { "data": encoding_spec['data'], "encoding": encoding_spec['encoding'] }
+              rest_client.post("/channels/#{channel_name}/messages", JSON.dump(raw_message))
+            end
+          end
+        end
+
+        context "when encoding #{encoding_spec['expectedType']}" do
+          it 'ensures that client libraries have compatible encoding and decoding using common fixtures' do
+            data = if encoding_spec['expectedHexValue']
+              encoding_spec['expectedHexValue'].scan(/../).map { |x| x.hex }.pack('c*')
+            else
+              encoding_spec['expectedValue']
+            end
+
+            realtime_channel.publish("event", data) do
+              response = rest_client.get("/channels/#{channel_name}/messages")
+              message = response.body[0]
+              expect(message['encoding']).to eql(encoding_spec['encoding'])
+              expect_data_to_match encoding_spec['data'], message['data']
```

Comment:
This is where order matters, as those values are JSON-encoded maps and can be like `{"a":123,"b":456}` and `{"b":456,"a":123}` and would fail to match given that they are `String`s. (Ruby seems to do what we expect, ie. to keep the order from the fixture, so no big deal I think.)